File: processing/src/main/java/org/apache/druid/query/DataSource.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.druid.query;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import org.apache.druid.java.util.common.Cacheable;
import org.apache.druid.query.planning.PreJoinableClause;
import org.apache.druid.query.policy.Policy;
import org.apache.druid.query.policy.PolicyEnforcer;
import org.apache.druid.segment.SegmentMapFunction;

import javax.annotation.Nullable;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Represents a source... of data... for a query. Analogous to the "FROM" clause in SQL.
 */
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type", defaultImpl = TableDataSource.class)
@JsonSubTypes({
    @JsonSubTypes.Type(value = TableDataSource.class, name = "table"),
    @JsonSubTypes.Type(value = QueryDataSource.class, name = "query"),
    @JsonSubTypes.Type(value = UnionDataSource.class, name = "union"),
    @JsonSubTypes.Type(value = JoinDataSource.class, name = "join"),
    @JsonSubTypes.Type(value = LookupDataSource.class, name = "lookup"),
    @JsonSubTypes.Type(value = InlineDataSource.class, name = "inline"),
    @JsonSubTypes.Type(value = GlobalTableDataSource.class, name = "globalTable"),
    @JsonSubTypes.Type(value = UnnestDataSource.class, name = "unnest"),
    @JsonSubTypes.Type(value = FilteredDataSource.class, name = "filter"),
    @JsonSubTypes.Type(value = RestrictedDataSource.class, name = "restrict")
})
public interface DataSource extends Cacheable
{
  byte JOIN_OPERATION_CACHE_ID = 0x1;
  byte TABLE_DATA_SOURCE_CACHE_ID = 0x2;
  byte NOOP_CACHE_ID = 0x71;

  /**
   * Returns the names of all table datasources involved in this query. Does not include names for non-tables, like
   * lookups or inline datasources.
   */
  Set<String> getTableNames();

  /**
   * Returns datasources that this datasource depends on. Will be empty for leaf datasources like 'table'.
   */
  List<DataSource> getChildren();

  /**
   * Return a new DataSource, identical to this one, with different children. The number of children must be equal
   * to the number of children that this datasource already has.
   */
  DataSource withChildren(List<DataSource> children);

  /**
   * Returns true if queries on this dataSource are cacheable at both the result level and per-segment level.
   * Currently, dataSources that do not actually reference segments (like 'inline'), are not cacheable since cache keys
   * are always based on segment identifiers.
   */
  boolean isCacheable(boolean isBroker);

  /**
   * Decides if this datasource can be accessed globally.
   * <p>
   * This means that all servers have a full copy of this datasource.
   * <p>
   * Examples: inline table, lookup.
   */
  boolean isGlobal();

  /**
   * Communicates that this {@link DataSource} can be directly used to run a {@link Query}.
   * <p>
   * A Processable datasource must pack the necessary logic into the {@link DataSource#createSegmentMapFunction(Query)}.
   * <p>
   * Processable examples are: {@link TableDataSource}, {@link InlineDataSource}, {@link FilteredDataSource}.
   * Non-processable ones are those which need further pre-processing before running them.
   * examples are: {@link QueryDataSource} and join which are not supported directly.
   */
  boolean isProcessable();

  /**
   * Returns a segment function on to how to segment should be modified.
   */
  SegmentMapFunction createSegmentMapFunction(Query query);

  /**
   * Returns an updated datasource based on the policy restrictions on tables.
   * <p>
   * If this datasource contains no table, no changes should occur.
   *
   * @param policyMap      a mapping of table names to policy restrictions. A missing key is different from an empty value:
   *                       <ul>
   *                         <li> a missing key means the table has never been permission checked.
   *                         <li> an empty value indicates the table doesn't have any policy restrictions, it has been permission checked.
   *                       </ul>
   * @param policyEnforcer the policy enforcer to enforce the result datasource complies
   * @return the updated datasource, with restrictions applied in the datasource tree
   * @throws IllegalStateException when mapping a RestrictedDataSource, unless the table has a NoRestrictionPolicy in
   *                               the policyMap (used by druid-internal). Missing policy or adding a
   *                               non-NoRestrictionPolicy to RestrictedDataSource would throw.
   */
  default DataSource withPolicies(Map<String, Optional<Policy>> policyMap, PolicyEnforcer policyEnforcer)
  {
    List<DataSource> children = this.getChildren()
                                    .stream()
                                    .map(child -> child.withPolicies(policyMap, policyEnforcer))
                                    .collect(Collectors.toList());
    return this.withChildren(children);
  }

  /**
   * Compute a cache key prefix for a data source. This includes the data sources that participate in the RHS of a
   * join as well as any query specific constructs associated with join data source such as base table filter. This key prefix
   * can be used in segment level cache or result level cache. The function can return following
   * - Non-empty byte array - If there is join datasource involved and caching is possible. The result includes
   * join condition expression, join type and cache key returned by joinable factory for each {@link PreJoinableClause}
   * - NULL - There is a join but caching is not possible. It may happen if one of the participating datasource
   * in the JOIN is not cacheable.
   *
   * @return the cache key to be used as part of query cache key
   */
  @Override
  @Nullable
  byte[] getCacheKey();
}

File: examples/quickstart/jupyter-notebooks/druidapi/datasource.py
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import requests, time
from .consts import COORD_BASE
from .rest import check_error
from .util import dict_get

REQ_DATASOURCES = COORD_BASE + '/datasources'
REQ_DATASOURCE = REQ_DATASOURCES + '/{}'

# Segment load status
REQ_DATASOURCES = COORD_BASE + '/datasources'
REQ_DS_LOAD_STATUS = REQ_DATASOURCES + '/{}/loadstatus'

class DatasourceClient:
    '''
    Client for status APIs. These APIs are available on all nodes.
    If used with the router, they report the status of just the router.
    '''
    
    def __init__(self, rest_client):
        self.client = rest_client

    def names(self, include_unused=False, include_disabled=False):
        """
        Returns a list of the names of data sources in the metadata store.
        
        Parameters
        ----------
        include_unused : bool, default = False
            if False, returns only datasources with at least one used segment
            in the cluster.

        include_unused : bool, default = False
            if False, returns only enamed datasources.

        Reference
        ---------
        * `GET /druid/coordinator/v1/metadata/datasources`
        * `GET /druid/coordinator/v1/metadata/datasources?includeUnused`
        * `GET /druid/coordinator/v1/metadata/datasources?includeDisabled`

        See https://druid.apache.org/docs/latest/operations/api-reference.html#get-4

        Note: this method uses a semi-deprecated API.
        See `Metadata.user_table_names()` for the preferred solution.
        """
        params = {}
        if include_unused:
            params['includeUnused'] = ''
        if include_disabled:
            params['includeDisabled'] = ''
        return self.client.get_json(REQ_DATASOURCES, params=params)
    
    def drop(self, ds_name, ifExists=False):
        """
        Drops a data source.

        Marks as unused all segments belonging to a datasource. 

        Marking all segments as unused is equivalent to dropping the table.
        
        Parameters
        ----------
        ds_name: str
            name of the datasource to query

        Returns
        -------
        Returns a map of the form 
        {"numChangedSegments": <number>} with the number of segments in the database whose 
        state has been changed (that is, the segments were marked as unused) as the result 
        of this API call.

        Reference
        ---------
        `DELETE /druid/coordinator/v1/datasources/{dataSourceName}`
        """
        r = self.client.delete(REQ_DATASOURCE, args=[ds_name])
        if ifExists and r.status_code == requests.codes.not_found:
            return
        check_error(r)

    def load_status_req(self, ds_name, params=None):
        return self.rest_client.get_json(REQ_DS_LOAD_STATUS, args=[ds_name], params=params)
    
    def load_status(self, ds_name):
        return self.load_status_req(ds_name, {
            'forceMetadataRefresh': 'true', 
            'interval': '1970-01-01/2999-01-01'})

    def wait_until_ready(self, ds_name):
        while True:
            resp = self.load_status(ds_name)
            if dict_get(resp, ds_name) == 100.0:
                return
            time.sleep(0.5)
            